Scripts/Helpers/ProxyUI.cs

```diff
@@ -0,0 +1,549 @@
+                        if (materialClone != null)
+                        {
+                            var originalAlpha = materialClone.GetFloat(shaderAlphaPropety);
+                            if (materialClone.HasProperty(k_ZWritePropertyName))
+                                materialClone.SetFloat(k_ZWritePropertyName, 1);
+
+                            m_BodyMaterialOriginalAlphaMap[materialClone] = new affordancePropertyTuple<float>(originalAlpha, originalAlpha);
+                        }
+                    }
+                    break;
+                case VisibilityControlType.materialSwap:
+                    m_BodySwapOriginalMaterials = new List<Material>();
+                    foreach (var renderer in m_BodyRenderers)
+                    {
+                        var materialClone = MaterialUtils.GetMaterialClone(renderer);
+                        m_BodySwapOriginalMaterials.Add(materialClone);
+                    }
+                    break;
+            }
+
+            affordancesVisible = false;
+            bodyVisible = false;
+
+            // Allow setting of affordance & body visibility after affordance+body setup is performed in the "affordances" property
+            m_ProxyUISetup = true;
+        }
+
+        IEnumerator AnimateAffordanceColorVisibility(bool isVisible, AffordanceDefinition definition)
+        {
+            const float kTargetAmount = 1.1f; // Overshoot in order to force the lerp to blend to maximum value, with needing to set again after while loop
+            var speedScalar = isVisible ? k_FadeInSpeedScalar : k_FadeOutSpeedScalar;
+            var currentAmount = 0f;
+            var visibilityDefinition = definition.visibilityDefinition;
+            var materialsAndColors = visibilityDefinition.materialsAndAssociatedColors;
+            var shaderColorPropety = visibilityDefinition.colorProperty;
+
+            if (materialsAndColors == null)
+                yield break;
+
+            // Setup animateFromColors using the current color values of each material associated with all renderers drawing this affordance
+            foreach (var materialAndAssociatedColors in materialsAndColors)
+            {
+                    var animateFromColor = materialAndAssociatedColors.firstElement.GetColor(shaderColorPropety); // Get current color from material
+                    var animateToColor = isVisible ? materialAndAssociatedColors.secondElement : materialAndAssociatedColors.thirdElement; // (second)original or (third)hidden color(alpha/color.a)
+                    materialAndAssociatedColors.fourthElement = animateFromColor;
+                    materialAndAssociatedColors.fifthElement = animateToColor;
+            }
+
+            while (currentAmount < kTargetAmount)
+            {
+                var smoothedAmount = MathUtilsExt.SmoothInOutLerpFloat(currentAmount += Time.unscaledDeltaTime * speedScalar);
```

Comment:
It's not as fancy but let's put the code that actually lets this loop complete on its own line